Source/WebCore/platform/graphics/cocoa/IOSurface.h

```diff
@@ -122,17 +122,17 @@ class IOSurface final {
 
     // Any images created from a surface need to be released before releasing
     // the surface, or an expensive GPU readback can result.
-    WEBCORE_EXPORT RetainPtr<CGImageRef> createImage();
-    WEBCORE_EXPORT static RetainPtr<CGImageRef> sinkIntoImage(std::unique_ptr<IOSurface>);
+    // Passed in context is the context through which the contents was drawn.
+    WEBCORE_EXPORT RetainPtr<CGImageRef> createImage(CGContextRef);
+    // Passed in context is the context through which the contents was drawn.
+    WEBCORE_EXPORT static RetainPtr<CGImageRef> sinkIntoImage(std::unique_ptr<IOSurface>, RetainPtr<CGContextRef>);
 
 #ifdef __OBJC__
     id asLayerContents() const { return (__bridge id)m_surface.get(); }
 #endif
     IOSurfaceRef surface() const { return m_surface.get(); }
-    WEBCORE_EXPORT CGContextRef ensurePlatformContext(PlatformDisplayID = 0);
-    // The graphics context cached on the surface counts as a "user", so to get
-    // an accurate result from isInUse(), it needs to be released.
-    WEBCORE_EXPORT void releasePlatformContext();
+
+    WEBCORE_EXPORT RetainPtr<CGContextRef> createPlatformContext(PlatformDisplayID = 0);
```

Comment:
What happens if multiple consumers create contexts for the same IOSurface and keep them alive (and both issue draw calls)?

Any created images are only going to guarantee to contain data from their respective context, and any drawing from another context may or may not (racily) be there.

It might be nice to only allow one platform context to be alive at any time, or at least add a comment warning about this?

Reply:
> What happens if multiple consumers create contexts for the same IOSurface and keep them alive (and both issue draw calls)?

The caller is expected to coordinate the mutation of the shared IOSurface. This is what we do when we read and write it both in GPUP and WP.

> It might be nice to only allow one platform context to be alive at any time, or at least add a comment warning about this?

I don't think it's possible / desirable in general. E.g. it's the same if you have a malloc buffer and would say it'd be nice if there would not be multiple CGBitmapContexts pointing to the same malloc buffer.

In other words: the code cannot be such that we would have random places creating random contexts to random IOSurfaces. Using IOSurfaces  needs to be intentional: who drew and with which and flushed when, why would the content be there, etc.